File: jifrozen/구현/자물쇠와_열쇠.java
class Solution {
    public static int n;

    // 2차원 배열 회전
    public static int[][] rotate(int[][] key) {
        int n = key.length;
        int result[][] = new int[n][n];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                result[n - 1 - j][i] = key[i][j];
            }
        }
        return result;
    }

    // 자물쇠 확인
    public static boolean check(int[][] newLock) {
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (newLock[n + i][n + j] != 1)
                    return false;
            }
        }
        return true;
    }

    public boolean solution(int[][] key, int[][] lock) {
        n = lock.length;
        int m = key.length;
        int[][] newLock = new int[n * 3][n * 3];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                newLock[i + n][j + n] = lock[i][j];
            }
        }

        for (int k = 0; k < 4; k++) {
            key = rotate(key);
            for (int a = 0; a < n * 3 - m; a++) {
                for (int b = 0; b < n * 3 - m; b++) {
                    for (int c = 0; c < m; c++) {
                        for (int d = 0; d < m; d++) {
                            newLock[a + c][b + d] += key[c][d];
                        }
                    }
                    if (check(newLock)) {
                        return true;
                    }
                    for (int c = 0; c < m; c++) {
                        for (int d = 0; d < m; d++) {
                            newLock[a + c][b + d] -= key[c][d];
                        }
                    }
                }
            }
        }
        return false;
    }
}

File: yunva17/구현/자물쇠와_열쇠.py


# 90도 회전
def turn(key):
    n = len(key)
    arr = [[0] * n for _ in range(n)]
    for i in range(n):
        for j in range(n):
            arr[j][n-i-1] = arr[i][j]

    return arr

# 자물쇠 키 확인
def check(lock):
    l = len(lock)//3
    for i in range(l, l*2):
        for j in range(l, l*2):
            if lock[i][j] != 1:
                return False
    return True


def solution(key, lock):
    a = len(lock)
    b = len(key)

    big = [[0]*3*a for _ in range(3*a)]

    for i in range(a):
        for j in range(a):
            big[i+a][j+a] = lock[i][j]

    
    for _ in range(4):
        key = turn(key)
        for x in range(a*2):
            for y in range(a*2):
                for i in range(b):
                    for j in range(b):
                        big[x+i][y+j] += key[i][j]
                
                if check(big) == True:
                    return True
                
                for i in range(b):
                    for j in range(b):
                        big[x+i][y+j] -= key[i][j]
    
    return False



